File: 361. Bomb Enemy.cpp
class Solution {
public:
    // DP: direction[i][j] -> how many enemt can be killed at this position from a specific direction
    // O(NM) time & space, 36.13%, ok (can reuse a same matrix for 4 directions, however, 4 matrixes are more understandable)
    int maxKilledEnemies(vector<vector<char>>& grid) {
        if(grid.empty()) return 0;
        int row = grid.size(), col = grid[0].size();
        int maxKill = 0;

        // from upperleft to lowerright
        vector<vector<int>> left(row, vector<int>(col, 0));
        vector<vector<int>> up(row, vector<int>(col, 0));
        for(int i = 0; i < row; i++){
            for(int j = 0; j < col; j++){
                if(grid[i][j] != 'W'){
                    up[i][j] = (i > 0 ? up[i - 1][j] : 0) + (grid[i][j] == 'E');
                    left[i][j] = (j > 0 ? left[i][j - 1] : 0) + (grid[i][j] == 'E');
                }
            }
        }

        // from lowerright to upperleft
        vector<vector<int>> right(row, vector<int>(col, 0));
        vector<vector<int>> down(row, vector<int>(col, 0));
        for(int i = row - 1; i >= 0; i--){
            for(int j = col - 1; j >= 0; j--){
                if(grid[i][j] != 'W'){
                    down[i][j] = (i < row - 1 ? down[i + 1][j] : 0) + (grid[i][j] == 'E');
                    right[i][j] = (j < col - 1 ? right[i][j + 1] : 0) + (grid[i][j] == 'E');
                    if(grid[i][j] != 'E')
                        maxKill = max(maxKill, up[i][j] + down[i][j] + left[i][j] + right[i][j]);
                }
            }
        }
        // printf("%d-%d-%d-%d", up[1][1], down[1][1], left[1][1], right[1][1]);
        return maxKill;
    }
};

File: 361. Bomb Enemy.py
class Solution(object):
    
#     # brute force: O(mn(m+n)) time, O(1) space, TLE
#     def maxKilledEnemies(self, grid):
#         """
#         :type grid: List[List[str]]
#         :rtype: int
#         """
#         if len(grid) == 0 or len(grid[0]) == 0:
#             return 0
#         m, n, ret = len(grid), len(grid[0]), 0   
#         def searchRow(left, right, row, step):
#             count = 0
#             for i in xrange(left, right, step):
#                 if grid[row][i] == 'W':
#                     return count
#                 if grid[row][i] == 'E':
#                     count += 1
#             return count
        
#         def searchColumn(up, down, col, step):
#             count = 0
#             for i in xrange(up, down, step):
#                 if grid[i][col] == 'W':
#                     return count
#                 if grid[i][col] == 'E':
#                     count += 1
#             return count
        
#         for i in xrange(m):
#             for j in xrange(n):
#                 if grid[i][j] == '0':
#                     cnt = searchColumn(i + 1, m, j, 1) + searchColumn(i - 1, -1, j, -1)
#                     cnt += searchRow(j + 1, n, i, 1) + searchRow(j - 1, -1, i, -1)
#                     ret = max(ret, cnt)
#         return ret

    # dynamic programming: O(mn) time, O(mn) space
    def maxKilledEnemies(self, grid):
        if len(grid) == 0 or len(grid[0]) == 0:
            return 0
        m, n, ret = len(grid), len(grid[0]), 0   
        [up, down, left, right] = [[[0 for j in range(n)] for i in range(m)] for k in range(4)]

        # upperleft to lowerright
        for i in xrange(m):
            for j in xrange(n):
                if grid[i][j] != 'W':
                    right[i][j] = int(grid[i][j] == 'E') + (right[i][j - 1] if j != 0 else 0)
                    down[i][j] = int(grid[i][j] == 'E') + (down[i - 1][j] if i != 0 else 0)
                    
        # lowerright to upperleft
        for i in xrange(m - 1, -1, -1):
            for j in xrange(n - 1, -1, -1):
                if grid[i][j] != 'W':
                    up[i][j] = int(grid[i][j] == 'E') + (up[i + 1][j] if i != m - 1 else 0)
                    left[i][j] = int(grid[i][j] == 'E') + (left[i][j + 1] if j != n - 1 else 0)
        
        for i in xrange(m):
            for j in xrange(n):
                if grid[i][j] == '0':
                    ret = max(ret, up[i][j] + down[i][j] + left[i][j] + right[i][j])
        return ret

    # better dynamic programming: O(mn) time, O(mn) space
    def maxKilledEnemies(self, grid):
        if len(grid) == 0 or len(grid[0]) == 0:
            return 0
        m, n, ret = len(grid), len(grid[0]), 0   
        [up, down, left, right] = [[[0 for j in range(n)] for i in range(m)] for k in range(4)]

        # upperleft to lowerright
        for i in xrange(m):
            for j in xrange(n):
                if grid[i][j] != 'W':
                    right[i][j] = int(grid[i][j] == 'E') + (right[i][j - 1] if j != 0 else 0)
                    down[i][j] = int(grid[i][j] == 'E') + (down[i - 1][j] if i != 0 else 0)
                    
        # lowerright to upperleft
        for i in xrange(m - 1, -1, -1):
            for j in xrange(n - 1, -1, -1):
                if grid[i][j] != 'W':
                    up[i][j] = int(grid[i][j] == 'E') + (up[i + 1][j] if i != m - 1 else 0)
                    left[i][j] = int(grid[i][j] == 'E') + (left[i][j + 1] if j != n - 1 else 0)
                if grid[i][j] == '0':  # move final check here 
                    ret = max(ret, up[i][j] + down[i][j] + left[i][j] + right[i][j])
        return ret
                    
        